### DISCUSSION TITLE
Bluetooth Low Energy (BLE) temperature / humidity beacon

### DISCUSSION THREAD

**Discussion by DavesCodeMusings** [micropython/micropython#15926] [Show and tell]
I recently got it into my head to DIY an ESP32-based Bluetooth Low Energy (BLE) temperature / humidity that I can put on my front porch and have it update Home Assistant with the weather conditions. Are there commercially available products that do this? Yes. Are there other DIY projects that do this? Probably. Are any of them as satisfying as coding it yourself with MicroPython's aioble library? Not on your life!

Anyway, the project is here if you're interested: https://github.com/DavesCodeMusings/BLESP32

What makes this different than the traditional, _Look Ma, I wrote my DHT22 readings to a GATT characteristic!_ is the use of the BLE Advertisement to broadcast the data. This lets the device blast out a few readings to Home Assistant and then go into deep sleep. Since it's not a GATT-based reading, Home Assistant doesn't seem to mind the device not being available for minutes on end. (And deep sleep helps mitigate self-heating throwing off the temperature readings as well.)

In itself, the project's not all that exciting. But, the fact that I now have a template for broadcasting whatever sensor data I want into my home automation system is pretty satisfying.

Below, you can see the temperature and humidity encoded in the BLE advertisement. A couple lambda functions in the home automation hub turn this into human readable information.

![broadcast_data](https://github.com/user-attachments/assets/16700984-f6d9-4710-b397-eed3df60edc6)

After spending hours messing with Arduino C++ to do something similar, I realized what a joy MicroPython really is.

**Comment by jonnor** [comment]
Nice project, thanks for sharing!
BLE advertisements are excellent for just dropping a couple of bytes of IoT sensor data from time to another, with very low power consumption. I also tested it with MicroPython recently, and found it to be quite easy (albeit not documented directly as a use-case). My test/example code here (using lightsleep not deepsleep, and low-level gap_advertise instead of aioble): https://github.com/jonnor/embeddedml/blob/master/handson/micropython-ble/ble_advertise_custom.py

**Reply by DavesCodeMusings** [reply]
I was pleasantly surprised at how quick it was to get this up and going with aioble. Particularly after trying it with C++ first. I started with the [aioble sensor example](https://github.com/micropython/micropython-lib/blob/master/micropython/bluetooth/aioble/examples/temp_sensor.py), tweaked a few things, and tossed the sensor readings into the _manufacturer data_ to feed to `aioble.advertise()`. Easy! I love MicroPython!!! Did I mention that?

Just curious, are you coding yours as part of a larger project or are you documenting some IoT howto examples? Like you, I didn't really find any direct examples associated with BLE beacons, which is part of my reason for sharing this code. Maybe there are more Home Assistant users with a bent on DIYing with MicroPython out there. Wouldn't be surprised.

**Reply by walklinewang** [reply]
You can also use Apple iBeacon payload to broadcast the advertising, just like below.

![b5fcfd6cc3a2a3ccd3f68eb3baa3b11](https://github.com/user-attachments/assets/72fd75f0-ce9a-4be1-bb31-74170349b1a2)

This is my test function,  just focus on the `_append` part of the code.

```python
def advertising_ibeacon_payload(proximity_uuid, major, minor, tx_power):
    """
    Generate paylaod for Apple iBeacon device
    """
    assert isinstance(proximity_uuid, UUID), TypeError("proximity uuid must be type of UUID")

    payload = bytearray()
    proximity_uuid = list(bytes(proximity_uuid))
    proximity_uuid.reverse()
    proximity_uuid = bytes(proximity_uuid)

    def _append(adv_type, value):
        nonlocal payload
        payload += PACK('BB', len(value) + 1, adv_type) + value

    _append(BLEConst.ADType.AD_TYPE_FLAGS, PACK('B', 0x06))
    _append(BLEConst.ADType.AD_TYPE_MANUFACTURER_SPECIFIC_DATA, PACK('<H2B', 0x004C, 0x02, 0x15) + proximity_uuid + PACK(">2HB", major, minor, tx_power))

    return payload
```

Apple iBeacon payload contains `major`, `minor` and `rssi` information.

**Reply by jonnor** [reply]
> curious, are you coding yours as part of a larger project or are you documenting some IoT howto examples?

Both kind-of :) At https://soundsensing.no we are now using MicroPython for prototyping additional/new sensors for monitoring machinery like HVAC systems. Our IoT commmunication system in production use primarily BLE advertisements, so it was useful to figure out a way to do that directly with MicroPython. We also sometimes send more raw data (larger payloads) when prototyping / testing, but that is then done over HTTPS.

I also research and develop TinyML (on device ML) solutions, including for MicroPython. Primarily organized around the project https://github.com/emlearn/emlearn-micropython 
In this case, I also do some feasibility studies, create demos and examples. For this, BLE advertisements are also relevant, since it's one of the best low-power solutions. And low-power is for some of the more serious use-cases, a key motivation for doing TinyML (as opposed to just sending the data to the cloud.

**Reply by DavesCodeMusings** [reply]
Thanks to both of you for providing coding examples. All my searching so far had only turned up either examples of how to do it with GATT characteristics or statements like, "Such and such device sends information in the advertisement, but it's proprietary and encrypted," with no leads on how to construct such an advertisement.

**Reply by walklinewang** [reply]
> with no leads on how to construct such an advertisement.

It seems to me that the purpose of construct advertising payload in a certain way is just for the purpose of "generality", e.g. words like `Apple` and `Nordic` can be recognised directly by some debugging app, which makes it easier for the user to accept it, and of course gives us a sense of achievement.

But essentially the broadcast payload is just 31 (and or more) bytes, and we can define the meaning of them ourselves, but then we lose the generality. 😶
